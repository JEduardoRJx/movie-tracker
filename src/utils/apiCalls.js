const baseUrl = 'https://api.themoviedb.org/3/movie/now_playing?';
// const apiKey = '1b20ae1afe685b2871c8d94218f89eba';
const apiKey2 = 'cee1e60becdb4297de68233fbef2f560';

export const fetchMovies = async () => {
  const url = `${baseUrl}api_key=${apiKey2}&language=en-US`
  const response = await fetch(url)
    if (!response.ok) {
      throw Error("There was an issue retrieving your movies. Please try again.");
    }
  let movies = await response.json();
  movies = movies.results.map(movie => {
    const { backdrop_path, genre_ids, id, overview, poster_path, release_date, title, vote_average } = movie
    return { backdrop_path, genre_ids, id, overview, poster_path, release_date, title, vote_average, isFavorite: false }
  })
  return movies
};

export const postUser = async user => {
  const url = 'http://localhost:3001/api/v1/login';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      email: user.email,
      password: user.password
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);
  if (!response.ok && response.status !== 401) {
    throw Error('Sorry, unable to retrieve your account. Try again later.')
  }
  const foundUser = await response.json();
  if (response.status === 401) {
    throw new Error('Username or password incorrect')
  }
  return foundUser;
}

export const addUser = async user => {
  const url = 'http://localhost:3001/api/v1/users';
  const options = {
    method: 'POST',
    body: JSON.stringify({
      name: user.name,
      email: user.email,
      password: user.password
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options);
    if (!response.ok && response.status !== 500) {
     throw Error('Sorry, unable to create your account. Try again later.')
  }
  const newUser = await response.json();
    if (response.status === 500) {
      throw new Error('There is already an account with this email. Go to login or use another email address.')
  }
  return newUser;
}

export const addFavorite = async (userId, movie) => {
  const url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites`;
  const options = {
    method: 'POST',
    body: JSON.stringify({
      movie_id: movie.id, 
      title: movie.title, 
      poster_path: movie.poster_path, 
      release_date: movie.release_date, 
      vote_average: movie.vote_average, 
      overview: movie.overview
    }),
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options)
  if(!response.ok) {
    throw Error('Unable to add movie as favorite. Try again later.')
  }
  const addedFav = await response.json();
  return addedFav;
}

export const getFavorites = async userId => {
  const url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites`;
  const response = await fetch(url)
  if(!response.ok) {
    throw Error('Unable to retrieve your favorite movies as this time. Please try again later')
  }
  const favoriteMovies = await response.json()
  return favoriteMovies
}

export const removeFavorite = async (userId, favId) => {
  const url = `http://localhost:3001/api/v1/users/${userId}/moviefavorites/${favId}`;
  const options = {
    method: 'DELETE',
    headers: {
      'Content-Type': 'application/json'
    }
  }
  const response = await fetch(url, options)
  if(!response.ok) {
    throw Error('Unable to remove this movie as a favorite at this time. Please try again later')
  }
}