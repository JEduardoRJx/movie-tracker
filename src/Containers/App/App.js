import React, { Component } from 'react';
import { Route } from "react-router-dom";
import { connect } from "react-redux";
import { fetchMovies, getFavorites, addFavorite, removeFavorite } from '../../utils/apiCalls';
import { addMovies, throwError, setFavs } from '../../actions';
import MoviesContainer from '../MoviesContainer/MoviesContainer';
import CardDetails from '../../components/CardDetails/CardDetails';
import CreateAccount from '../CreateAccount/CreateAccount'
import LoginForm from '../LoginForm/LoginForm';
import Header from '../Header/Header';
import './App.scss';
// import PropTypes from 'prop-types';

class App extends Component {

  async componentDidMount() {
    const { addMovies, throwError } = this.props
    try {
      const movies = await fetchMovies();
      addMovies(movies)
    } catch({ message }) {
      throwError(message)
    }
  }

  refreshFavs = async (movie) => {
    const { throwError, setFavs, user } = this.props;
    try {
      if (movie.isFavorite) {
        await removeFavorite(user.id, movie.id)
      } else {
        await addFavorite(user.id, movie)
      }
        const updatedFavs = await getFavorites(user.id)
        setFavs(updatedFavs)
      } catch ({ message }) {
        throwError(message)
      }
  } 

  render() {
    const { errorMsg, user } = this.props
    console.log("user in app render", user)
    return (
      <div className="App">
        <Header />
          {errorMsg && <h2>{errorMsg}</h2>}
          <Route exact path='/' render={() => <MoviesContainer viewAll={true} refreshFavs={this.refreshFavs}/>}/>
          <Route exact path='/login' 
            render={() => <LoginForm />} />
          <Route exact path='/createaccount' 
            render={() => <CreateAccount />} />
          <Route exact path='/favorites' render={() => <MoviesContainer viewAll={false} refreshFavs={this.refreshFavs}/>}/>
          <Route
            path="/movies/:id"
            render={({ match }) => {
              let foundMovie = this.props.movies.find(movie => {
                return parseInt(movie.id) == match.params.id;
              });
              return (
                <CardDetails
                  key={foundMovie.id}
                  movie={foundMovie}
                  refreshFavs={this.refreshFavs}
              />
            );
          }}
        />
      </div>
    );
  }
}

const mapStateToProps = ({ errorMsg, user, movies }) => ({
  errorMsg,
  user,
  movies
})

const mapDispatchToProps = dispatch => ({
  addMovies: movies => dispatch(addMovies(movies)),
  throwError: errorMsg => dispatch(throwError(errorMsg)),
  setFavs: favs => dispatch(setFavs(favs)) 
})

export default connect(mapStateToProps, mapDispatchToProps)(App);
