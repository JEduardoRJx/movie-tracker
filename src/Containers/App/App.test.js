import React from 'react';
import { shallow } from 'enzyme';
import { App } from './App';
import { fetchMovies, getFavorites, addFavorite} from '../../utils/apiCalls';
jest.mock('../../utils/apiCalls');

describe('App', () => {
  let wrapper;
  const mockAddMovies = jest.fn();
  const mockThrowError = jest.fn();
  const mockSetFavs = jest.fn();
  const mockUser = {
    id: 2, 
    name: "Alex", 
    email: "alex@gmail.com"
  };
  const mockMovies = [{
    backdrop_path: "/n6bUvigpRFqSwmPp1m2YADdbRBc.jpg",
    genre_ids: (3) [80, 18, 53],
    id: 475557,
    isFavorite: false,
    overview: "During the 1980s, a failed stand-up comedian is driven insane and turns to a life of crime and chaos in Gotham City while becoming an infamous psychopathic crime figure.",
    poster_path: "/udDclJoHjfjb8Ekgsd4FDteOkCU.jpg",
    release_date: "2019-10-04",
    title: "Joker",
    vote_average: 8.6,
    isFavorite: false
  }];
  const mockFaves = [mockMovies[0]];

  fetchMovies.mockImplementation(() => Promise.resolve(mockMovies));
  addFavorite.mockImplementation(() => Promise.resolve(mockUser.id, mockMovies[0]));
  getFavorites.mockImplementation(() => Promise.resolve(mockFaves));

  beforeEach(() => {
    wrapper = shallow(<App 
      addMovies={mockAddMovies} 
      throwError={mockThrowError} 
      errorMsg={''} 
      user={mockUser} 
      movies={mockMovies} 
      setFavs={mockSetFavs}
      />);
  });

  it('should match snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should set movies when fetch movies is called', () => {
    expect(fetchMovies).toHaveBeenCalled();
    expect(mockAddMovies).toHaveBeenCalledWith(mockMovies);
  });
  
  it('should be able to favorite a movie and get favorites', async () => {
    await wrapper.instance().refreshFavs(mockMovies[0]);
    expect(addFavorite).toHaveBeenCalledWith(mockUser.id, mockMovies[0]);
    expect(getFavorites).toHaveBeenCalledWith(2);
    expect(mockSetFavs).toHaveBeenCalledWith(mockFaves);
  });
});