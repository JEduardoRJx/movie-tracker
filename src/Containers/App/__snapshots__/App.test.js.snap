// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match snapshot 1`] = `
<div
  className="App"
>
  <Connect(Header) />
  <Route
    exact={true}
    path="/"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/login"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/createaccount"
    render={[Function]}
  />
  <Route
    exact={true}
    path="/favorites"
    render={[Function]}
  />
  <Route
    path="/movies/:id"
    render={[Function]}
  />
</div>
`;
